content/2017.02.handling-python-projects-dependencies.rst

```diff
@@ -0,0 +1,225 @@
+=====================================
+
+You might be tempted to pin dependencies version in your setup.py
+
+Because you are using a contrain file you are safe for future updates
+that might break your code.
+
+On your CI not using a contrain file will help you to detect that you
+need to take some actions to support the new released version.
+
+There are some cases when you still want to pin some version:
+
+
+If you know that your project will not work with a lower version
+----------------------------------------------------------------
+
+If you are using a feature or API that didn't existed before or you
+had a bug that was fixed after.
+
+.. code-block::
+
+   psycopg > 2.5
+   colander >= 1.3.2
+   ujson >= 1.35
+
+
+If you know that your project doesn't support yet the next release
+------------------------------------------------------------------
+
+**I insist that this implies that the release already exists and that
+your test suite doesn't run on it.**
+
+In that case, and only in that case, you can do it for the shorted
+possible time until you ported your project to it.
+
+It is often the case for frameworks:
+
+.. code-block::
+
+   Pyramid < 1.8
+   django >1.6,<= 1.8
+
+The danger of doing it is that you migh create conflict and in that
+case Python will not let you start your application.
+
+An easy way to break things is to ping a max ``requests`` version for instance:
+
+.. code-block::
+
+   requests < 2.13
+
+If you do that you will end up having a lot of
+``pkg_resources.VersionConflict`` error when running your program
+because Python will check the dependency and will not accept to run if
+you have a greater ``requests`` version which can happen if another
+dependency already needed ``requests`` as a dependency and pip already
+installed it.
+
+**Really don't do it**
+
+
+What about test dependencies?
+=============================
+
+That's a good question, I am glad you asked.
+
+Test dependencies are less of an issue. You can either use
+``test_requires`` in your ``setup.py`` or a ``dev-requirements.txt``
+file.
+
+In the former you will need to run tests using ``python setup.py test``
+which unfortunately doesn't run in a ``virtualenv``
+
+In the later you will need to make sure your test dependencies are
+installed before running the tests but tools like ``tox`` already do
+that for you.
+
+
+Our take on this is to put test dependencies in a
+``dev-requirements.txt`` file.
+
+We have a Makefile rule that knows if dev-dependencies should be
+installed or not before running the ``tests`` target.
+
+As a bonus itt will automatically create a virtualenv if you don't
+have one already activated:
+
+.. code-block:: Makefile
+
+    VIRTUALENV = virtualenv --python=python3
+	VENV := $(shell echo $${VIRTUAL_ENV-.venv})  # Use the activated virtualenv path or use .venv
+	PYTHON = $(VENV)/bin/python
+	INSTALL_STAMP = $(VENV)/.install.stamp
+	INSTALL_DEV_STAMP = $(VENV)/.dev_env_installed.stamp
+
+	install: $(INSTALL_STAMP)
```

Comment:
I'm not an expert on Makefiles, but I think 195-201 should all be indented the same.